lib/internal/fs/recursive_watch.js

```diff
@@ -0,0 +1,290 @@
+'use strict';
+
+const {
+  ArrayPrototypePush,
+  SafePromiseAllReturnVoid,
+  Promise,
+  PromisePrototypeThen,
+  SafeMap,
+  StringPrototypeStartsWith,
+  StringPrototypeSlice,
+  StringPrototypeLastIndexOf,
+  SymbolAsyncIterator,
+} = primordials;
+
+const { EventEmitter } = require('events');
+const assert = require('internal/assert');
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { getValidatedPath } = require('internal/fs/utils');
+const { kFSWatchStart, StatWatcher } = require('internal/fs/watchers');
+const { kEmptyObject } = require('internal/util');
+const { validateBoolean, validateAbortSignal } = require('internal/validators');
+const path = require('path');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = new SafeMap()) {
+  const { opendir } = lazyLoadFsPromises();
+
+  const filenames = await opendir(dir);
```

Comment:
```suggestion
  const filenames = opendir(dir);
```
since `for await`

Reply:
I don't think that's true, promises are not async iterable.

Reply:
@aduh95 is right.

```
TypeError: filenames is not async iterable
    at traverse (node:internal/fs/recursive_watch:48:28)
    at [kFSWatchStart] (node:internal/fs/recursive_watch:227:11)
    at Object.watch (node:fs:2307:36)
    at /home/parallels/Desktop/node/test/parallel/test-fs-watch-recursive-symlink.js:30:22
    at Object.<anonymous> (/home/parallels/Desktop/node/test/parallel/test-fs-watch-recursive-symlink.js:51:3)
    at Module._compile (node:internal/modules/cjs/loader:1175:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1229:10)
    at Module.load (node:internal/modules/cjs/loader:1038:32)
    at Module._load (node:internal/modules/cjs/loader:879:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:82:12)
```

Reply:
my bad. but then why use `for await` instead of just `for`